### PR TITLE
UART HAL

### PR DESCRIPTION
**Pull request by AdityaAsGithub**

### CODE REVIEW

File: common/phal_F4_F7/gpio/gpio.h [hunk 1]
```diff
@@ -177,16 +177,26 @@ typedef struct
...
 #define GPIO_INIT_SPI2_SCK_PB13 GPIO_INIT_AF(GPIOB, 13, 5, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_PUSH_PULL, GPIO_INPUT_PULL_DOWN)
 #define GPIO_INIT_SPI2_MISO_PB14 GPIO_INIT_AF(GPIOB, 14, 5, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_OPEN_DRAIN, GPIO_INPUT_OPEN_DRAIN)
 #define GPIO_INIT_SPI2_MOSI_PB15 GPIO_INIT_AF(GPIOB, 15, 5, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_PUSH_PULL, GPIO_INPUT_PULL_DOWN)
+
 #define GPIO_INIT_SPI1_SCK_PB3 GPIO_INIT_AF(GPIOB, 3, 5, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_PUSH_PULL, GPIO_INPUT_PULL_DOWN)
 #define GPIO_INIT_SPI1_MOSI_PB5 GPIO_INIT_AF(GPIOB, 5, 5, GPIO_OUTPUT_HIGH_SPEED, GPIO_OUTPUT_PUSH_PULL, GPIO_INPUT_PULL_DOWN)
+
+
+//UART/USART
 #define GPIO_INIT_USART3TX_PC10 GPIO_INIT_AF(GPIOC, 10, 7, GPIO_OUTPUT_ULTRA_SPEED, GPIO_TYPE_AF, GPIO_INPUT_OPEN_DRAIN)
```
Comment (LukeOxley) [inline common/phal_F4_F7/gpio/gpio.h:194]: the GPIO_TYPE_AF does not go there, it should be a GPIO_OUTPUT pull type. Same goes for other UARTs. Could also fix it in the CAN tx and rx defs too.

File: common/phal_F4_F7/usart/usart.h [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#define USART2_ACTIVE_IDX 1
+#define USART3_ACTIVE_IDX 2
+#define USART4_ACTIVE_IDX 3
+#define USART5_ACTIVE_IDX 4
+#define USART6_ACTIVE_IDX 5
+#define USART7_ACTIVE_IDX 6
+#define USART8_ACTIVE_IDX 7
+
+// Defines that mean the same thing but are phrased differently in stm32f4xx and stm32f7xx
+#define USART_BRR_DIV_Fraction_Pos USART_BRR_DIV_FRACTION_Pos
```
Comment (LukeOxley) [inline common/phal_F4_F7/usart/usart.h:44]: lol

File: common/phal_F4_F7/usart/usart.h [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+{
+    OB_DISABLE, //!< Disable one bit sampling
+    OB_ENABLE   //!< Enable one bit sampling
+} obsample_t;
+
+typedef enum
+{
+    USART_DMA_TX,   //!< USART is transmitting over DMA
+    USART_DMA_RX    //!< USART is recieving over DMA
+} usart_dma_mode_t;
```
Comment (LukeOxley) [inline common/phal_F4_F7/usart/usart.h:102]: Is this mutually exclusive?

Comment (LukeOxley) [reply common/phal_F4_F7/usart/usart.h:102]: Maybe make this inside source file if it's only meant to be used in there.

File: common/phal_F4_F7/usart/usart.h [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    volatile usart_rx_errors_t rx_errors; //!< And RX error flags set during reception
+} usart_init_t;
+
+typedef struct {
+    usart_init_t *active_handle;   //!< USART handle provided on initialization
+    uint8_t cont_rx;               //!< Flag controlling RX rececption mode (once or continously)
+    uint8_t _tx_busy;              //!< Waiting on a transmission to finish
+    volatile uint8_t _rx_busy;     //!< Waiting on a reception to finish
+    volatile uint32_t rxfer_size;  //!< Size of data to receive over DMA
+} usart_active_transfer_t;
```
Comment (LukeOxley) [inline common/phal_F4_F7/usart/usart.h:155]: maybe put this in source file as well

File: common/phal_F4_F7/usart/usart.h [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+/**
+ * @brief Callback function called immediately after reception of a USART RX message
+ * Uses USART IDLE line interrupt
+ *
+ * NOTE: this is executed during an interrupt handler call, so keep code in this function light
+ *
+ * @param handle Handle of USART peripheral that just recieved a message
+ */
+extern void usart_recieve_complete_callback(usart_init_t *handle);
```
Comment (LukeOxley) [inline common/phal_F4_F7/usart/usart.h:222]: If this works it will greatly simplify parsing

File: common/phal_F4_F7/usart/usart.h [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    bool overrun;        //!< USART unable to parse data in time
+    bool noise_detected; //!< Oversampling detected a possible bit flip due to noise in usart frame
+    bool framing_error; //!< Unable to understand USART frame
+    bool parity_error; //!< USART Parity bit incorrect (Only when parity is enabled)
+
+    // DMA Errors
+    bool dma_transfer_error; //!< DMA transfer error
+    bool dma_direct_mode_error; //!< DMA error while attempting to operate in direct mode
+    bool dma_fifo_overrun; //!< DMA FIFO has been overrun - apparently this can be ignored on USART peripherals AS LONG AS YOU AREN'T USING THE FIFO
+} usart_rx_errors_t;
```
Comment (LukeOxley) [inline common/phal_F4_F7/usart/usart.h:115]: All the comments r v helpful

**Review by LukeOxley — CHANGES_REQUESTED** [review]
couple teeny things
overall looking really good

File: common/phal_F4_F7/usart/usart.h (hunk 5 above)
Comment (AdityaAsGithub) [reply common/phal_F4_F7/usart/usart.h:222]: It does!

**Review by LukeOxley — APPROVED** [review]
LGTM

**Pull request merged**